business/istio_certs.go

```diff
@@ -0,0 +1,143 @@
+	Certificates []struct {
+		SecretName string   `yaml:"secretName"`
+		DNSNames   []string `yaml:"dnsNames"`
+	} `yaml:"certificates"`
+}
+
+type CertInfo struct {
+	SecretName string    `json:"secretName"`
+	DNSNames   []string  `json:"dnsNames"`
+	Issuer     string    `json:"issuer"`
+	Subject    string    `json:"subject"`
+	NotBefore  time.Time `json:"notBefore"`
+	NotAfter   time.Time `json:"notAfter"`
+	Error      error     `json:"error"`
+}
+
+func (ics *IstioCertsService) GetCertsInfo() ([]CertInfo, error) {
+	if certs, err := ics.getChironCertificates(); err == nil {
+		return certs, nil
+	} else if cert, err := ics.getCertificateFromSecret("cacerts"); err == nil {
+		return []CertInfo{cert}, nil
+	} else {
+		cert, err := ics.getCertificateFromSecret("istio-ca-secret")
+		if err != nil {
+			return nil, err
+		}
+		return []CertInfo{cert}, nil
+	}
+}
+
+func (ics *IstioCertsService) getCertificateFromSecret(secretName string) (CertInfo, error) {
+	cfg := config.Get()
+
+	secret, err := ics.k8s.GetSecret(cfg.IstioNamespace, secretName)
+
+	if err != nil {
+		log.Println(err)
+		return CertInfo{}, err
+	}
+	log.Println(string(secret.Data["ca-cert.pem"]))
+	block, _ := pem.Decode(secret.Data["ca-cert.pem"])
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return CertInfo{SecretName: secretName, Error: err}, nil
+	}
+	log.Println(cert.NotBefore)
+	log.Println(cert.NotAfter)
+	return CertInfo{
+		Issuer:     cert.Issuer.String(),
+		Subject:    cert.Subject.String(),
+		NotBefore:  cert.NotBefore,
+		NotAfter:   cert.NotAfter,
+		SecretName: secretName,
+	}, nil
+}
+
+func (ics *IstioCertsService) getChironCertificates() ([]CertInfo, error) {
+	cfg := config.Get()
+
+	istioConfigMap, err := ics.k8s.GetConfigMap(cfg.IstioNamespace, "istio")
+	if err != nil {
+		return nil, err
+	}
+
+	istioConfig := IstioConfig{}
+	err = yaml.Unmarshal([]byte(istioConfigMap.Data["mesh"]), &istioConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if istioConfig.Certificates == nil {
+		return nil, errors.New("no certificates configured in istio configmap")
+	}
+
+	wg := sync.WaitGroup{}
+	certChan := make(chan CertInfo, len(istioConfig.Certificates))
+	errChan := make(chan error, len(istioConfig.Certificates))
+
+	for _, certConfig := range istioConfig.Certificates {
+		wg.Add(1)
+		go func(secretName string, dnsNames []string) {
+			defer wg.Done()
+			certSecret, err := ics.k8s.GetSecret(cfg.IstioNamespace, secretName)
+			if err != nil {
+				errChan <- err
+				return
+			} else {
+				block, _ := pem.Decode(certSecret.Data["cert-chain.pem"])
+				cert, err := x509.ParseCertificate(block.Bytes)
+				if err != nil {
+					certChan <- CertInfo{SecretName: secretName, Error: err}
+					return
+				}
+				certChan <- CertInfo{
+					Issuer:     cert.Issuer.String(),
+					Subject:    cert.Subject.String(),
+					NotBefore:  cert.NotBefore,
+					NotAfter:   cert.NotAfter,
+					SecretName: secretName,
+					DNSNames:   dnsNames,
+				}
```

Comment:
I think you can reuse this piece of code somehow. Less error-prone for modifications.

Reply:
True, will do that.